src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -480,9 +720,361 @@ private partial class KeyValueListNode
+        W3C,          // 00-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX-XXXXXXXXXXXXXXXX-XX see https://w3c.github.io/trace-context/
+    };
+
+    /// <summary>
+    /// A TraceId is the format the W3C standard requires for its ID for the entire trace. 
+    /// It represents 16 binary bytes of information, typically displayed as 32 characters
+    /// of Hexadecimal.  A TraceId is a STRUCT, and does contain the 16 bytes of binary information
+    /// so there is value in passing it by reference.   It does know how to convert to and
+    /// from its Hexadecimal string representation, tries to avoid changing formats until
+    /// it has to, and caches the string representation after it was created.   
+    /// It is mostly useful as an exchange type.  
+    /// </summary>
+    public unsafe struct ActivityTraceId : IEquatable<ActivityTraceId>
+    {
+        /// <summary>
+        /// Creates a TraceId from a 16 byte span 'idBytes'.   The bytes are copied.  
+        /// </summary>
+        public ActivityTraceId(ReadOnlySpan<byte> idData, bool isUtf8Chars = false)
+        {
+            _id1 = 0;
+            _id2 = 0;
+            _asHexString = null;
+            fixed (ulong* idPtr = &_id1)
+            {
+                var outBuff = new Span<byte>(idPtr, sizeof(ulong) * 2);
+                if (isUtf8Chars)
+                {
+                    if (idData.Length != 32)
+                        throw new ArgumentOutOfRangeException(nameof(idData));
+                    ActivityTraceId.SetSpanFromHexUtf8Chars(outBuff, idData);
+                }
+                else
+                {
+                    if (idData.Length != 16)
+                        throw new ArgumentOutOfRangeException(nameof(idData));
+                    idData.CopyTo(outBuff);
+                }
+            }
+        }
+
+        public ActivityTraceId(ReadOnlySpan<char> idData)
+        {
+            _id1 = 0;
+            _id2 = 0;
+            _asHexString = null;
+            fixed (ulong* idPtr = &_id1)
+            {
+                if (idData.Length != 32)
+                    throw new ArgumentOutOfRangeException(nameof(idData));
+                ActivityTraceId.SetSpanFromHexChars(new Span<byte>(idPtr, sizeof(ulong) * 2), idData);
+            }
+        }
+
+        /// <summary>
+        /// Create a new TraceId with at random number in it (very likely to be unique)
+        /// </summary>
+        public static ActivityTraceId NewTraceId()
+        {
+            ActivityTraceId ret = new ActivityTraceId();
+            SetToRandomBytes(new Span<byte>(&ret._id1, sizeof(ulong) * 2));
+            return ret;
+        }
+
+        /// <summary>
+        /// Return the bytes of the ID.  Its length is always 16
+        /// </summary>
+        public ReadOnlySpan<byte> AsBytes
+        {
+            get
+            {
+                fixed (ulong* idPtr = &_id1)
+                    return new ReadOnlySpan<byte>(idPtr, sizeof(ulong) * 2);
+            }
```

Comment:
This looks dangerous. What if this ActivityTraceId is stored on a class?  The GC could move the instance in memory, thus also moving the location of the ActivityTraceId and its _id1 field, and now this ReadOnlySpan is pointing to the wrong place.

Reply:
AsBytes removed replaced with CopyTo